src/edu/stanford/dlss/was/WasapiDownloader.java

```diff
@@ -44,10 +46,81 @@ void downloadSelectedWarcs() throws IOException {
       WasapiCrawlSelector crawlSelector = new WasapiCrawlSelector(wasapiRespList);
       for (Integer crawlId : desiredCrawlIds(crawlSelector)) {
         for (WasapiFile file : crawlSelector.getFilesForCrawl(crawlId)) {
-          // TODO:  make a separate method for downloading individual file?
-          System.out.println("We will eventually download " + file.toString());
+          downloadAndValidateFile(file);
+        }
+      }
+    }
+  }
+
+  // package level method for testing
+  WasapiConnection getWasapiConn() throws IOException {
+    if (wasapiConn == null)
+      wasapiConn = new WasapiConnection(new WasapiClient(settings));
+    return wasapiConn;
+  }
+
+  // package level method for testing
+  @SuppressWarnings("checkstyle:MethodLength")
+  void downloadAndValidateFile(WasapiFile file) throws NoSuchAlgorithmException {
+    String fullFilePath = prepareOutputLocation(file);
+    if (fullFilePath == null) {
+      System.err.println("fullFilePath is null - can't retrieve file");
+      return;
+    }
```

Comment:
i'm slightly uncomfortable with this, just because i can't quite see how that method could ever return a `null`... is it something that fails in the tests?  if so, could the call to `prepareOutputLocation` be intercepted so it can return a mock value?

but also, if it works, it doesn't seem like that big of a deal, and a warning does get printed.

or, totally possible i'm misunderstanding `prepareOutputLocation()`.